File: tests/typescript/eks-existing-cluster.test.ts
/**
 * Mock Tests for Connecting Existing EKS Clusters to CAST AI (TypeScript)
 *
 * These tests simulate connecting an existing EKS cluster to CAST AI for optimization.
 * The cluster already exists - we're just onboarding it to CAST AI.
 *
 * Run with: npm test
 */

import * as pulumi from "@pulumi/pulumi";
import * as castai from "@castai/pulumi";
import { promisify, promisifyAll } from "./test-utils";

/**
 * Mock implementation for CAST AI and AWS resources.
 */
class CastAIMocks implements pulumi.runtime.Mocks {
    newResource(args: pulumi.runtime.MockResourceArgs): { id: string; state: any } {
        const outputs = { ...args.inputs };

        if (args.type === "castai:aws:EksCluster") {
            return {
                id: `${args.name}-cluster-id-${this.hash(args.name)}`,
                state: {
                    ...outputs,
                    id: `${args.name}-cluster-id-${this.hash(args.name)}`,
                    agentToken: `mock-eks-token-${this.hash(args.name)}`,
                    credentialsId: `mock-credentials-${this.hash(args.name)}`,
                },
            };
        }

        return {
            id: `${args.name}-id`,
            state: outputs,
        };
    }

    call(args: pulumi.runtime.MockCallArgs): Record<string, any> {
        return {};
    }

    private hash(str: string): number {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            const char = str.charCodeAt(i);
            hash = (hash << 5) - hash + char;
            hash = hash & hash;
        }
        return Math.abs(hash) % 1000;
    }
}

// Set up mocks
pulumi.runtime.setMocks(new CastAIMocks());

describe("EKS Existing Cluster Connection", () => {
    it("should connect an existing EKS cluster to CAST AI", async () => {
        const cluster = new castai.EksCluster("existing-eks-cluster", {
            accountId: "123456789012",
            region: "us-east-1",
            name: "production-eks-cluster", // Existing cluster name
            deleteNodesOnDisconnect: false, // Don't delete nodes when disconnecting
            overrideSecurityGroups: ["sg-existing123"],
            subnets: ["subnet-existing-a", "subnet-existing-b"],
        });

        const [clusterId, clusterName, agentToken, accountId] = await promisifyAll(
            cluster.id,
            cluster.name,
            cluster.agentToken,
            cluster.accountId
        );

        expect(clusterId).toBeDefined();
        expect(clusterName).toBe("production-eks-cluster");
        expect(agentToken).toBeDefined();
        expect(accountId).toBe("123456789012");
    });

    it("should connect existing EKS cluster using IAM assume role", async () => {
        const assumeRoleArn = "arn:aws:iam::123456789012:role/CastAI-CrossAccountRole";

        const cluster = new castai.EksCluster("existing-eks-assume-role", {
            accountId: "123456789012",
            region: "us-west-2",
            name: "staging-eks-cluster",
            deleteNodesOnDisconnect: false,
            overrideSecurityGroups: ["sg-existing456"],
            subnets: ["subnet-existing-1", "subnet-existing-2"],
            assumeRoleArn: assumeRoleArn,
        });

        const [roleArn, clusterName] = await promisifyAll(
            cluster.assumeRoleArn,
            cluster.name
        );

        expect(roleArn).toBe(assumeRoleArn);
        expect(clusterName).toBe("staging-eks-cluster");
    });

    it("should connect existing EKS cluster with SSH public key", async () => {
        const sshPublicKey = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQ... existing-key";

        const cluster = new castai.EksCluster("existing-eks-ssh", {
            accountId: "123456789012",
            region: "eu-west-1",
            name: "dev-eks-cluster",
            deleteNodesOnDisconnect: true,
            overrideSecurityGroups: ["sg-dev-123"],
            subnets: ["subnet-dev-a", "subnet-dev-b", "subnet-dev-c"],
            sshPublicKey: sshPublicKey,
        });

        const [sshKey, clusterName] = await promisifyAll(
            cluster.sshPublicKey,
            cluster.name
        );

        expect(sshKey).toBe(sshPublicKey);
        expect(clusterName).toBe("dev-eks-cluster");
    });

    it("should connect existing EKS cluster with resource tags", async () => {
        const clusterTags = {
            Environment: "production",
            ManagedBy: "castai",
            Team: "platform",
            CostCenter: "engineering",
        };

        const cluster = new castai.EksCluster("existing-eks-tags", {
            accountId: "987654321098",
            region: "us-east-2",
            name: "prod-eks-main",
            deleteNodesOnDisconnect: false,
            overrideSecurityGroups: ["sg-prod-main"],
            subnets: ["subnet-prod-1a", "subnet-prod-1b"],
            tags: clusterTags,
        });

        const [tags, clusterName] = await promisifyAll(cluster.tags, cluster.name);

        expect(tags).toBeDefined();
        if (tags) {
            expect(Object.keys(tags).length).toBe(4);
            expect(tags.Environment).toBe("production");
            expect(tags.Team).toBe("platform");
        }
        expect(clusterName).toBe("prod-eks-main");
    });

    it("should connect existing EKS cluster with minimal configuration", async () => {
        const cluster = new castai.EksCluster("existing-eks-minimal", {
            accountId: "111111111111",
            region: "ap-southeast-1",
            name: "minimal-eks",
            deleteNodesOnDisconnect: true,
            overrideSecurityGroups: ["sg-minimal"],
            subnets: ["subnet-min-a"],
        });

        const [clusterId, clusterName, accountId] = await promisifyAll(
            cluster.id,
            cluster.name,
            cluster.accountId
        );

        expect(clusterId).toBeDefined();
        expect(clusterName).toBe("minimal-eks");
        expect(accountId).toBe("111111111111");
    });

    it("should connect existing EKS cluster with custom DNS configuration", async () => {
        const customDnsIp = "172.20.0.10";

        const cluster = new castai.EksCluster("existing-eks-dns", {
            accountId: "222222222222",
            region: "us-west-2",
            name: "custom-dns-cluster",
            deleteNodesOnDisconnect: false,
            overrideSecurityGroups: ["sg-dns-123"],
            subnets: ["subnet-dns-1", "subnet-dns-2"],
            dnsClusterIp: customDnsIp,
        });

        const [dnsIp, clusterName] = await promisifyAll(
            cluster.dnsClusterIp,
            cluster.name
        );

        expect(dnsIp).toBe(customDnsIp);
        expect(clusterName).toBe("custom-dns-cluster");
    });

    it("should handle delete_nodes_on_disconnect for production clusters", async () => {
        // Production: keep nodes when disconnecting (safety)
        const prodCluster = new castai.EksCluster("existing-eks-prod", {
            accountId: "333333333333",
            region: "us-east-1",
            name: "prod-critical-cluster",
            deleteNodesOnDisconnect: false, // IMPORTANT: Don't delete in production
            overrideSecurityGroups: ["sg-prod"],
            subnets: ["subnet-prod-a", "subnet-prod-b"],
        });

        // Dev: can delete nodes when disconnecting
        const devCluster = new castai.EksCluster("existing-eks-dev", {
            accountId: "333333333333",
            region: "us-west-2",
            name: "dev-test-cluster",
            deleteNodesOnDisconnect: true, // Safe for dev environments
            overrideSecurityGroups: ["sg-dev"],
            subnets: ["subnet-dev-a"],
        });

        const [prodDelete, devDelete] = await promisifyAll(
            prodCluster.deleteNodesOnDisconnect,
            devCluster.deleteNodesOnDisconnect
        );

        expect(prodDelete).toBe(false);
        expect(devDelete).toBe(true);
    });

    it("should connect multiple existing EKS clusters across regions", async () => {
        const regions = ["us-east-1", "us-west-2", "eu-west-1"];
        const clusters = regions.map(
            (region, i) =>
                new castai.EksCluster(`existing-eks-${region}`, {
                    accountId: "444444444444",
                    region: region,
                    name: `cluster-${region}`,
                    deleteNodesOnDisconnect: false,
                    overrideSecurityGroups: ["sg-multi"],
                    subnets: ["subnet-multi"],
                })
        );

        const clusterRegions = await Promise.all(
            clusters.map((c) => promisify(c.region))
        );

        clusterRegions.forEach((region: string, i: number) => {
            expect(region).toBe(regions[i]);
        });
    });
});

File: tests/go/eks_existing_cluster_test.go
package tests

import (
	"testing"

	"github.com/castai/pulumi-castai/sdk/go/castai"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/stretchr/testify/assert"
)

// Tests for connecting existing EKS clusters to CAST AI

func TestConnectExistingEksCluster(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Connect existing cluster - in real scenario, cluster already exists in AWS
		cluster, err := castai.NewEksCluster(ctx, "existing-eks-cluster", &castai.EksClusterArgs{
			AccountId:               pulumi.String("123456789012"),
			Region:                  pulumi.String("us-east-1"),
			Name:                    pulumi.String("production-eks-cluster"), // Existing cluster name
			DeleteNodesOnDisconnect: pulumi.Bool(false),                     // Don't delete nodes when disconnecting
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-existing123"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-existing-a"),
				pulumi.String("subnet-existing-b"),
			},
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksWithAssumeRole(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Real-world scenario: Using cross-account assume role
		assumeRoleArn := "arn:aws:iam::123456789012:role/CastAI-CrossAccountRole"

		cluster, err := castai.NewEksCluster(ctx, "existing-eks-assume-role", &castai.EksClusterArgs{
			AccountId:               pulumi.String("123456789012"),
			Region:                  pulumi.String("us-west-2"),
			Name:                    pulumi.String("staging-eks-cluster"),
			DeleteNodesOnDisconnect: pulumi.Bool(false),
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-existing456"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-existing-1"),
				pulumi.String("subnet-existing-2"),
			},
			AssumeRoleArn: pulumi.String(assumeRoleArn), // Use IAM role instead of access keys
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksWithSSHKey(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		sshPublicKey := "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQ... existing-key"

		cluster, err := castai.NewEksCluster(ctx, "existing-eks-ssh", &castai.EksClusterArgs{
			AccountId:               pulumi.String("123456789012"),
			Region:                  pulumi.String("eu-west-1"),
			Name:                    pulumi.String("dev-eks-cluster"),
			DeleteNodesOnDisconnect: pulumi.Bool(true),
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-dev-123"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-dev-a"),
				pulumi.String("subnet-dev-b"),
				pulumi.String("subnet-dev-c"),
			},
			SshPublicKey: pulumi.String(sshPublicKey),
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksWithTags(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		clusterTags := pulumi.StringMap{
			"Environment": pulumi.String("production"),
			"ManagedBy":   pulumi.String("castai"),
			"Team":        pulumi.String("platform"),
			"CostCenter":  pulumi.String("engineering"),
		}

		cluster, err := castai.NewEksCluster(ctx, "existing-eks-tags", &castai.EksClusterArgs{
			AccountId:               pulumi.String("987654321098"),
			Region:                  pulumi.String("us-east-2"),
			Name:                    pulumi.String("prod-eks-main"),
			DeleteNodesOnDisconnect: pulumi.Bool(false),
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-prod-main"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-prod-1a"),
				pulumi.String("subnet-prod-1b"),
			},
			Tags: clusterTags,
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksMinimalConfig(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Minimal config - just the required fields
		cluster, err := castai.NewEksCluster(ctx, "existing-eks-minimal", &castai.EksClusterArgs{
			AccountId:               pulumi.String("111111111111"),
			Region:                  pulumi.String("ap-southeast-1"),
			Name:                    pulumi.String("minimal-eks"),
			DeleteNodesOnDisconnect: pulumi.Bool(true),
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-minimal"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-min-a"),
			},
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksCustomDNS(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Some clusters use custom DNS cluster IP
		customDnsIp := "172.20.0.10"

		cluster, err := castai.NewEksCluster(ctx, "existing-eks-dns", &castai.EksClusterArgs{
			AccountId:               pulumi.String("222222222222"),
			Region:                  pulumi.String("us-west-2"),
			Name:                    pulumi.String("custom-dns-cluster"),
			DeleteNodesOnDisconnect: pulumi.Bool(false),
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-dns-123"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-dns-1"),
				pulumi.String("subnet-dns-2"),
			},
			DnsClusterIp: pulumi.String(customDnsIp),
		})
		assert.NoError(t, err)
		assert.NotNil(t, cluster)
		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectExistingEksProductionVsDev(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		// Production: keep nodes when disconnecting (safety)
		prodCluster, err := castai.NewEksCluster(ctx, "existing-eks-prod", &castai.EksClusterArgs{
			AccountId:               pulumi.String("333333333333"),
			Region:                  pulumi.String("us-east-1"),
			Name:                    pulumi.String("prod-critical-cluster"),
			DeleteNodesOnDisconnect: pulumi.Bool(false), // IMPORTANT: Don't delete in production
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-prod"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-prod-a"),
				pulumi.String("subnet-prod-b"),
			},
		})
		assert.NoError(t, err)
		assert.NotNil(t, prodCluster)

		// Dev: can delete nodes when disconnecting
		devCluster, err := castai.NewEksCluster(ctx, "existing-eks-dev", &castai.EksClusterArgs{
			AccountId:               pulumi.String("333333333333"),
			Region:                  pulumi.String("us-west-2"),
			Name:                    pulumi.String("dev-test-cluster"),
			DeleteNodesOnDisconnect: pulumi.Bool(true), // Safe for dev environments
			OverrideSecurityGroups: pulumi.StringArray{
				pulumi.String("sg-dev"),
			},
			Subnets: pulumi.StringArray{
				pulumi.String("subnet-dev-a"),
			},
		})
		assert.NoError(t, err)
		assert.NotNil(t, devCluster)

		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}

func TestConnectMultipleExistingEksClusters(t *testing.T) {
	err := pulumi.RunErr(func(ctx *pulumi.Context) error {
		regions := []string{"us-east-1", "us-west-2", "eu-west-1"}

		for i, region := range regions {
			cluster, err := castai.NewEksCluster(ctx, "existing-eks-"+region, &castai.EksClusterArgs{
				AccountId:               pulumi.String("444444444444"),
				Region:                  pulumi.String(region),
				Name:                    pulumi.String("cluster-" + region),
				DeleteNodesOnDisconnect: pulumi.Bool(false),
				OverrideSecurityGroups: pulumi.StringArray{
					pulumi.String("sg-multi"),
				},
				Subnets: pulumi.StringArray{
					pulumi.String("subnet-multi"),
				},
			})
			assert.NoError(t, err)
			assert.NotNil(t, cluster, "Cluster %d should not be nil", i)
		}

		return nil
	}, pulumi.WithMocks("project", "stack", &CastAIMocks{}))

	assert.NoError(t, err)
}
